src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -915,21 +906,137 @@ public override void ReportDiagnosticsIfObsolete(Binder binder, SyntaxNode synta
+        {
+            private readonly TypeParameterSymbol _typeParameter;
+            private readonly bool? _isNullableIfReferenceType;
+
+            internal LazyNonNullableTypeParameter(TypeParameterSymbol typeParameter, bool? isNullableIfReferenceType)
+            {
+                Debug.Assert(isNullableIfReferenceType != true); // Should use LazyNullableTypeParameter.
+                _typeParameter = typeParameter;
+                _isNullableIfReferenceType = isNullableIfReferenceType;
+            }
+
+            public override TypeSymbol TypeSymbol => _typeParameter;
+            public override bool? IsNullable =>  _isNullableIfReferenceType == false || !_typeParameter.IsReferenceType ? (bool?)false : null;
+
+            public override ImmutableArray<CustomModifier> CustomModifiers => ImmutableArray<CustomModifier>.Empty;
+
+            public override TypeSymbolWithAnnotations AsNotNullableReferenceType()
+            {
+                return _isNullableIfReferenceType == false ?
+                    this :
+                    new LazyNonNullableTypeParameter(_typeParameter, isNullableIfReferenceType: false);
+            }
+
+            public override TypeSymbolWithAnnotations AsNullableReferenceType()
+            {
+                return new LazyNullableTypeParameter(_typeParameter.ContainingAssembly, this);
+            }
+
+            public override TypeSymbolWithAnnotations AsObliviousReferenceType()
+            {
+                return _isNullableIfReferenceType == null ?
+                    this :
+                    new LazyNonNullableTypeParameter(_typeParameter, isNullableIfReferenceType: null);
+            }
+
+            public override TypeSymbol AsTypeSymbolOnly()
+            {
+                throw new NotImplementedException();
+            }
 
+            public override bool Is(TypeParameterSymbol other)
+            {
+                return _typeParameter == other;
+            }
+
+            public override TypeSymbolWithAnnotations SubstituteType(AbstractTypeMap typeMap)
+            {
+                var newUnderlying = typeMap.SubstituteType(_typeParameter);
+                if ((object)newUnderlying != this)
+                {
+                    if ((newUnderlying.TypeSymbol.Equals(_typeParameter, TypeCompareKind.AllAspects) ||
+                            newUnderlying.TypeSymbol is IndexedTypeParameterSymbolForOverriding) &&
+                        newUnderlying.CustomModifiers.IsEmpty)
+                    {
+                        return newUnderlying;
+                    }
+
+                    return base.SubstituteType(typeMap);
+                }
+                else
+                {
+                    return this; // substitution had no effect on the type or modifiers
+                }
+            }
+
+            public override TypeSymbolWithAnnotations SubstituteTypeWithTupleUnification(AbstractTypeMap typeMap)
```

Comment:
Is this implementation identical to that in `LazyNullableTypeParameter`?
It's also very similar to `SubstituteType` above. Maybe "with tuple unification" should be a option to pass into `SubstituteType`?

Maybe worth adding a PROTOTYPE marker to streamline.